### PR TITLE
Introducing CardDAV into core

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@karlitschek @jancborchardt @PVince81 @schiesbn @blizzz @icewind1991 @MorrisJobke please review and test

@Henni here we go!

### CODE REVIEW

**Comment by karlitschek** [comment]
nice. As discussed :+1:  Can we automatically migrate the data from the calender/contacts backend?

**Comment by DeepDiver1975** [comment]
> Can we automatically migrate the data from the calender/contacts backend?

We should do this via vcard export from the current app and then import in here.
I'm still making up my mind about the best apporach:
- document it properly
- blog about it
- write a tool - questionable since we need a web based approach as well
- ...

**Comment by schiesbn** [comment]
Awesome, that's great!

> We should do this via vcard export from the current app and then import in here.

I agree, the contacts app already support the export so we just need a import to the new backend. But we should probably do this only if contacts uses the new backend and/or gets replaced by a different contacts app. If we migrate the address books earlier it doesn't make much sense because it will get out of sync again if people continue to use the contacts app.

Just some brainstorming, I see several ideas how this could be done:

- contacts switches to the new backend: In this case we could implement it as a upgrade step executed for each user separately the first time he log in after the update. On a per user basis it shouldn't be to expensive to do it during first login.
- a general background job which does it for all users, but then we would need to block the contacts app for each user until his address book was migrated -> could be disturbing for the users if they have to wait and don't know when it will be ready.
- a new contacts app: This app could have a button "export from address book from old contacts app" which every user can execute individually.

**Comment by DeepDiver1975** [comment]
> If we migrate the address books earlier it doesn't make much sense because it will get out of sync again if people continue to use the contacts app.

It definitely has to be an explicit step. Any automagic will cause confusion here.



> a new contacts app: This app could have a button "export from address book from old contacts app" which every user can execute individually.

That could be a valid approach @Henni

**Comment by DeepDiver1975** [comment]
There is some escape issue with the IN operator

@nickvergessen any idea? THX

````
Tests\Connector\Sabre\CardDavBackendTest::testMultiCard
Doctrine\DBAL\Exception\InvalidFieldNameException: An exception occurred while executing 'SELECT `id`, `uri`, `lastmodified`, `etag`, `size`, `carddata` FROM `oc_cards` WHERE (`addressbookid` = ?) AND (`uri` IN (`cardfIrhaWQHVuPse`, `cardPmb9U0qHlXLaS`))' with params ["3"]:

SQLSTATE[42S22]: Column not found: 1054 Unknown column 'cardfIrhaWQHVuPse' in 'where clause'

**Comment by karlitschek** [comment]
Hmmm. Maybe I'm the only guy here with a different idea. ;-) Automatic is better most of the time. Or can we have a 'migrate' button that pulls the vcard from contacts and imports them? 

A crazy idea. Could we even reuse the existing DB table?
What do you think? @jancborchardt

**Comment by jancborchardt** [comment]
I’d also say automatic migration (and backup in case of error) is better than asking people to manually click buttons. Would be really cool if that’s possible because then it’s totally seamless. If it’s not, some people might not even consider upgrading or be bothered.

**Comment by schiesbn** [comment]
@karlitschek 

> Or can we have a 'migrate' button that pulls the vcard from contacts and imports them? 

I think that's what @DeepDiver1975 and I understand if we say "manually" with a "export from address book from old contacts app" button :wink: 

@Henni

> That could be a valid approach @Henni

sounds like you volunteer to work on a (new) contacts app? :smile:

**Comment by Henni** [comment]
I'm already on it: Henni/contactsrework#3
It isn't in a working state yet, but it will be soon.

**Comment by DeepDiver1975** [comment]
> I think that's what @DeepDiver1975 and I understand if we say "manually" with a "export from address book from old contacts app" button :wink:

Exactly - we offer the user a button to export from the old app and import it to the new app.
This is something we might want to track in Henni/contactsrework

@Henni I guess it's time to name the app bit more serios :wink:

**Comment by DeepDiver1975** [comment]
Failing OCI unit tests are unrelated

File: apps/dav/tests/unit/carddav/carddavbackendtest.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+namespace Tests\Connector\Sabre;
```
Comment (nickvergessen) [inline apps/dav/tests/unit/carddav/carddavbackendtest.php]: OCA\DAV\...

**Comment by nickvergessen** [comment]
@DeepDiver1975 
> There is some escape issue with the IN operator

This is resolved?

**Comment by DeepDiver1975** [comment]
> This is resolved?

yes - thx

File: apps/dav/tests/unit/carddav/carddavbackendtest.php (hunk 1 above)
Comment (MorrisJobke) [inline apps/dav/tests/unit/carddav/carddavbackendtest.php]: @DeepDiver1975 Please fix this ;)

**Comment by jancborchardt** [comment]
@Henni let’s work on that app at the next meetup most definitely! Or are you also up for some hacking in the next days? I’m back in Munich.

**Comment by schiesbn** [comment]
> I'm already on it: Henni/contactsrework#3
> It isn't in a working state yet, but it will be soon.

That's awesome! I'm really looking forward to it! :smile:

**Comment by DeepDiver1975** [comment]
I'd really love to get this in - please connect your carddav client and let me know if it works - THX

@schiesbn @blizzz @PVince81 @georgehrke @LukasReschke @karlitschek THX

**Comment by LukasReschke** [comment]
I'd happy to test this and thus I do risk a silly question: What is the endpoint URL? :speak_no_evil:

**Comment by blizzz** [comment]
I don't know… with cadaver i can login an browse addressbooks. When I login as alice, i can go into the addressbooks/admin/ for instance.

However, Kontact refuses to use anything. When I try to set it up and look for collections, querying /master/remote.php/dav/ results in a 401, followed by an authenticated request resulting in 207, but Kontact does not offer me any collections to pick. Same if I provide /master/remote.php/dav/addressbooks/user/ as URL.

The CardDAV setup account in my phone is happy setting this up, but when I try to sync the server is contacted, but without authentication data. However, 400 is returned. The request was against remote.php/dav/, but the ends up the same way going against remote.php/dav/addressbooks/user/

Difference between Kontact and my phones CardDAV client and cadaver is that cadaver operates with OPTIONS first while the others do PROPFIND.

I don't know who is to blame: @DeepDiver1975's code, both my GUIful clients, or I… :chicken:

**Comment by DeepDiver1975** [comment]
> I'd happy to test this and thus I do risk a silly question: What is the endpoint URL? :speak_no_evil:

remote.php/dav is the root of all dav resources
remote.php/dav/calendars/${user} will point you to all the users calendars.

**Comment by DeepDiver1975** [comment]
@blizzz hmm - thanks for testing - I have it properly running with Gnome Evolution ...

I'll give it a second look. THX

**Comment by DeepDiver1975** [comment]
Carddav sync on Android is not working :imp:

**Comment by DeepDiver1975** [comment]
Gnome Evolution works if there is a calendar existing.  We need a way to create an addressbook.

occ command  for now until we have a client which can take care of this?

@Henni please add a 'new addressbook' button ;-)

**Comment by blizzz** [comment]
Yesterday I only tried with addressbooks since the title says CardDAV. Am confused about Calendar mentions.

**Comment by DeepDiver1975** [comment]
> Yesterday I only tried with addressbooks since the title says CardDAV. Am confused about Calendar mentions.

Yeah - sorry for the confusion - this PR is about CardDAV only.
I'm confusing myself all day long and mix up these ... :wink:

**Comment by blizzz** [comment]
All fine then ;)

After creating an addressbook, Kontact still does not offer any collection. 

With my mobile's cardDAV I still get back only 400. 

Tried those paths with both clients:
```
/remote.php/dav/
/remote.php/dav/addressbooks/$user/
/remote.php/dav/addressbooks/$user/$addressbook
```

**Comment by DeepDiver1975** [comment]
Looks like something is missing to detect the resources properly

**Comment by DeepDiver1975** [comment]
@blizzz maybe kde requires the .welknown file to be properly setup? afaik this is required on mac as well

**Comment by blizzz** [comment]
@DeepDiver1975 maybe, however i have not seen in the server log that any client was requesting it.

**Comment by georgehrke** [comment]
Adding an account in OS X doesn't work. Neither with address book nor without one.
Code looks good though

**Comment by DeepDiver1975** [comment]
THX Georg for testing. I have to read up how all these clients do the initial request.

**Comment by georgehrke** [comment]
Do we plan to support not having an address book or are we going to introduce some mechanism to automatically create one if there isn't?

**Comment by DeepDiver1975** [comment]
I just moved sharing to a PR of it's own - https://github.com/owncloud/core/pull/20355

**Comment by DeepDiver1975** [comment]
> Do we plan to support not having an address book or are we going to introduce some mechanism to automatically create one if there isn't?

@jancborchardt ???

**Comment by Henni** [comment]
@DeepDiver1975 usability-wise I would prepopulate the database with a single empty addressbook, so that it is immediately ready to use.

Maybe we could even add a first contact containing information about the user (name, e-mail, ...).

**Comment by DeepDiver1975** [comment]
> Maybe we could even add a first contact containing information about the user (name, e-mail, ...).

there is the me-card concept for the own contact - http://sabre.io/dav/carddav-me-card/

**Comment by DeepDiver1975** [comment]
Let me add the default calendar ...

@jancborchardt it's name? 'My Contacts' ?

**Comment by georgehrke** [comment]
... and what if a user purposely deletes their address book?
We see that some clients break and users will blame us.

**Comment by DeepDiver1975** [comment]
I added the default calendar/addressbook topic to the meta issue - https://github.com/owncloud/core/issues/20047

**Comment by DeepDiver1975** [comment]
regarding the missing lookup capabilities to allow various clients to connect:
I contacted Marten Gajda - the developer of CardDAV-Sync to get this issue sorted out.

From my pov we should move on and get this sorted out in followup issues/prs.

Objections? @PVince81 @georgehrke @schiesbn @Henni @blizzz THX

**Comment by georgehrke** [comment]
> Objections?

Not from my side.

**Comment by PVince81** [comment]
Follow up PR is fine, make sure to have a ticket so we don't forget :smile:

**Comment by blizzz** [comment]
okay

**Comment by jancborchardt** [comment]
Folks, just look at the current Contacts app:
* one empty address book by default of course
* name: »Contacts«

Additionally it should not be possible to delete it, unless there are other address books. There always needs to be one address book for Contacts to be useful.

**Comment by rullzer** [comment]
When I login with cadaver I see all users in addressbooks. However they seem to have a timestamp of 0. Maybe that is the reason some clients refuse to sync?

**Comment by DeepDiver1975** [comment]
> Maybe that is the reason some clients refuse to sync?

according to my observations regarding CardDAV-sync andoird app - the basic connectivity is failing.
So the client does not even detect the addressbooks .... well let's see

**Comment by Henni** [comment]
How do I test this correctly? 
cadaver doesn't work for me
```
$ cadaver http://localhost:8080/remote.php/dav/ 
Could not open collection:
Could not parse response status line
```
and the command
```
$ curl -X PROPFIND http://testuser:testpasswd@localhost:8080/remote.php/dav/
```
returns an html response which looks like this:
image

any idea @DeepDiver1975 ?

**Comment by DeepDiver1975** [comment]
@Henni anything in the logs?

**Comment by Henni** [comment]
this was a local issue, which is now resolved.
Looks good to me then!

**Comment by DeepDiver1975** [comment]
> regarding the missing lookup capabilities to allow various clients to connect:
> I contacted Marten Gajda - the developer of CardDAV-Sync to get this issue sorted out.

**Comment by DeepDiver1975** [comment]
toooo stupid to press the right buttons .... :imp:

**Comment by dmfs** [comment]
@DeepDiver1975 do you have any log from the CardDAV-Sync account setup that contains an error?

**Comment by DeepDiver1975** [comment]
> do you have any log from the CardDAV-Sync account setup that contains an error?

No .. mind pointing me to the docs? :wink:

**Comment by dmfs** [comment]
When the account setup fails you should see an error message and a "show log" button that opens the log with an option to send it via email. If the "wait ..." message keeps spinning "forever", there is probably an issue at TCP/IP level.

Am 6. November 2015 17:23:53 MEZ, schrieb "Thomas Müller" <notifications@github.com>:
>> do you have any log from the CardDAV-Sync account setup that contains
>an error?
>
>No .. mind pointing me to the docs? :wink: 
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/owncloud/core/pull/20184#issuecomment-154457744

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by DeepDiver1975** [comment]
> When the account setup fails you should see an error message and a "show log" button that opens the log with an option to send it via email. If the "wait ..." message keeps spinning "forever", there is probably an issue at TCP/IP level.

THX - I did not see any 'show log' button - let me have another look ...

**Comment by DeepDiver1975** [comment]
no show log button.

here is the apache access log:
````
192.168.178.50 - - [06/Nov/2015:17:38:09 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 401 1230 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - admin [06/Nov/2015:17:38:09 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 207 1029 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /oc/core/remote.php/ HTTP/1.1" - 8451 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /.well-known/carddav HTTP/1.1" 405 581 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /SOGo/dav/ HTTP/1.1" 405 571 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /groupdav.php/ HTTP/1.1" 404 508 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /dav/ HTTP/1.1" 405 566 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /carddav/ HTTP/1.1" 405 570 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /principals/ HTTP/1.1" 405 573 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND / HTTP/1.1" 405 572 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /remote.php/carddav/ HTTP/1.1" 404 514 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /owncloud/remote.php/carddav/ HTTP/1.1" 405 590 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:10 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 401 1233 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - admin [06/Nov/2015:17:38:10 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 207 903 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /oc/core/remote.php/ HTTP/1.1" - 8455 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /.well-known/carddav/ HTTP/1.1" 405 582 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /SOGo/dav/ HTTP/1.1" 405 571 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /groupdav.php/ HTTP/1.1" 404 508 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /dav/ HTTP/1.1" 405 566 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /carddav/ HTTP/1.1" 405 570 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /principals/ HTTP/1.1" 405 573 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND / HTTP/1.1" 405 572 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /remote.php/carddav/ HTTP/1.1" 404 514 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /owncloud/remote.php/carddav/ HTTP/1.1" 405 590 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 401 1231 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - admin [06/Nov/2015:17:38:11 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 207 1029 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - - [06/Nov/2015:17:38:11 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 401 1231 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"
192.168.178.50 - admin [06/Nov/2015:17:38:12 +0100] "PROPFIND /oc/core/remote.php/dav/ HTTP/1.1" 207 1029 "-" "CardDAV-Sync free/0.4.14 (motorola; osprey_reteu; Android 5.1.1; de_DE; org.dmfs.carddav.sync/125)"

**Comment by DeepDiver1975** [comment]
ci is green and we have enough quotes -> merging

**Comment by dmfs** [comment]
@DeepDiver1975 The missing "show log" option sounds like a known issue which will be fixed in the next release. If the server responded with a valid multistatus but no address books were found, it doesn't show any error and no "show log" button.

The Apache log indicates that there may be an issue with the first server response. With the first request the app tries to find the user's principal. Since the next request goes to `/oc/core/remote.php/` I think that's the principal address returned by the server. Otherwise the app wouldn't have accessed that URL. There is no status code logged for this request, but it must have been some kind of error, because the app continues with the standard discovery procedure.

Do you have this version running on some public IP?

**Comment by dmfs** [comment]
btw. does this merge request mean, that the CalDAV/CardDAV path names will change in an upcoming release? If so, will there be a redirect from the old URLs or do users need to reconfigure their clients?

**Comment by karlitschek** [comment]
redirects/migration is needed. yes

**Comment by dmfs** [comment]
would be great if you could ship it with an .htaccess file that rewrites the old path names to the new ones by default then.

**Comment by karlitschek** [comment]
.htaccess is not always respected by the webserver. So keeping the old endpoints with explicit redirects might be better

**Comment by dmfs** [comment]
yeah, true. Sounds like you've already planned to provide that kind of backwards compatibility, which is great.

**Comment by DeepDiver1975** [comment]
> Do you have this version running on some public IP?

I can set this up during the weekend

**Comment by dmfs** [comment]
@DeepDiver1975 Thanks a lot!

It looks like the server doesn't know the DAV property `DAV:current-user-principal` anymore (see [RFC 5397](https://tools.ietf.org/html/rfc5397))

Our client first tries to find the current-user-principal by sending the following `PROPFIND` request to the entered address (with the header `depth: 0`):

```xml
<?xml version="1.0" encoding="UTF-8"?>
<A:propfind xmlns:A="DAV:">
  <A:prop>
    <A:current-user-principal/>
    <A:resourcetype/>
  </A:prop>
</A:propfind>
```

OC 9 responds with 
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <d:response>
    <d:href>/remote.php/dav/</d:href>
    <d:propstat>
      <d:prop>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
    <d:propstat>
      <d:prop>
        <d:current-user-principal/>
      </d:prop>
      <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

It says `<d:current-user-principal/>` is `NOT FOUND`.

Compare the result of an OC 8 instance (the request was sent to `/remote.php/carddav`):
```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
  <d:response>
    <d:href>/remote.php/carddav/</d:href>
    <d:propstat>
      <d:prop>
        <d:current-user-principal>
          <d:href>/remote.php/carddav/principals/USERNAME/</d:href>
        </d:current-user-principal>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

Even if I send the request to the actual principal path (i.e. `/remote.php/dav/principals/USERNAME/`) I get a broken response. The principal URL must return the resource type `DAV:principal` (see [RFC 3744](https://tools.ietf.org/html/rfc3744#section-4).

This is the OC 9 response:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <d:response>
    <d:href>/remote.php/dav/principals/USERNAME/</d:href>
    <d:propstat>
      <d:prop>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
    <d:propstat>
      <d:prop>
        <d:current-user-principal/>
      </d:prop>
      <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

As you can see the `DAV:principal` resource type is not present.

Again, compare it to the response of an OC 8 server:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
  <d:response>
    <d:href>/remote.php/carddav/principals/USERNAME/</d:href>
    <d:propstat>
      <d:prop>
        <d:current-user-principal>
          <d:href>/remote.php/carddav/principals/USERNAME/</d:href>
        </d:current-user-principal>
        <d:resourcetype>
          <d:collection/>
          <d:principal/>
        </d:resourcetype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

As you can see the OC 8 response contains `<d:principal/>` in the resource type element.

**Comment by dmfs** [comment]
There is more. The property `CARDDAV:addressbook-home-set` is unknown too (see https://tools.ietf.org/html/rfc6352#section-7.1.1). This property should be defined on the principal URL.

Here is the request I send to `/remote.php/dav/principals/USERNAME/`:

```xml
<?xml version="1.0" encoding="utf-8" ?>
<A:propfind xmlns:A="DAV:" xmlns:B="urn:ietf:params:xml:ns:carddav">
  <A:prop>
    <A:displayname/>
    <A:resourcetype/>
    <B:addressbook-home-set/>
  </A:prop>
</A:propfind>
```

Here is what OC 9 returns:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <d:response>
    <d:href>/remote.php/dav/principals/USERNAME/</d:href>
    <d:propstat>
      <d:prop>
        <d:displayname>USERNAME</d:displayname>
        <d:resourcetype>
          <d:collection/>
        </d:resourcetype>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
    <d:propstat>
      <d:prop>
        <x2:addressbook-home-set xmlns:x2="urn:ietf:params:xml:ns:carddav"/>
      </d:prop>
      <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

No `CARDDAV:addressbook-home-set`.

Again, here is what OC 8 returns:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
  <d:response>
    <d:href>/remote.php/carddav/principals/USERNAME/</d:href>
    <d:propstat>
      <d:prop>
        <d:displayname>USERNAME</d:displayname>
        <d:resourcetype>
          <d:collection/>
          <d:principal/>
        </d:resourcetype>
        <card:addressbook-home-set>
          <d:href>/remote.php/carddav/addressbooks/USERNAME/</d:href>
        </card:addressbook-home-set>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```
It points to the right addressbook home set.

The same problem exists with `CALDAV:calendar-home-set`.

**Comment by dmfs** [comment]
FYI you can use `curl` to send the requests from a shell like so:

```shell
curl -k -X PROPFIND -H "content-type: application/xml" -u USERNAME:PASSWORD \
--data-binary '<?xml version="1.0" encoding="UTF-8"?><A:propfind xmlns:A="DAV:"><A:prop><A:current-user-principal/><A:resourcetype/></A:prop></A:propfind>' \
-H "Depth: 0" https://SERVER/remote.php/dav/ | xmllint --format -
```

Just replace `USERNAME`, `PASSWORD` and `SERVER` with the actual values. I've piped the output through `xmllint` to prettify the response XML.

To send the addressbook-home-set query use:

```shell
curl -k -X PROPFIND -H "content-type: application/xml" -u USERNAME:PASSWORD \
--data-binary '<?xml version="1.0" encoding="utf-8" ?><A:propfind xmlns:A="DAV:" xmlns:B="urn:ietf:params:xml:ns:carddav"><A:prop><A:displayname/><A:resourcetype/><B:addressbook-home-set/></A:prop></A:propfind>' \
-H "Depth: 0" https://SERVER/remote.php/dav/principals/USERNAME/ | xmllint --format -
```

**Comment by DeepDiver1975** [comment]
THX a lot @dmfs - I'll fix this asap

**Pull request merged**